### PR TITLE
#385 drools update WIP

### PR DESCRIPTION
**Pull request by mwolfthal**
#385 Changes related to phase 1 of Drools update refactoring. Most of the information in the individual commits is in docs/DroolsUpdate.md.

### CODE REVIEW

**Comment by jasonaowen** [comment]
This approach is looking good! Thank you, @mwolfthal.

I do have some requests on structuring commits. As we've been working on the PSM over the last year and a half, we've tried to build up a detailed and useful commit history, so that tools like `git blame` (and their IDE equivalents) are useful, and so that future developers can understand the reasoning behind the choices we make.

To that end, I'd like to request both that you break up some commits, and that you merge others. In particular, some of the commits (like 4057643cd4980f2bbfe03a84daf5b8cd386d7afe) have several different *kinds* of changes in them: splitting long lines, moving initializations, changing interfaces, adding JavaDoc, and renaming variables; it would make it easier to review (and easier for future developers to understand) if each of those were separate commits. Likewise, some of the changes that are being made are to move methods between classes; each such move should be all in the same commit, rather than split across multiple commits.

Also in that vein, we would like to [reference issues](https://github.com/SolutionGuidance/psm/blob/master/CONTRIBUTING.md#reference-issues) in commits that are related to those issues; this allows people who are looking at the history to better understand the context and the motivation of a commit. Please make sure that each of these commits includes the number and the title of the relevant issue(s).

Does that make sense? I'm happy to talk about this more!

File: psm-app/cms-business-process/src/main/java/gov/medicaid/domain/rules/DroolsKnowledgeDelegate.java [hunk 1]
```diff
@@ -40,13 +40,15 @@
 import org.jbpm.task.service.local.LocalTaskService;
 
 /**
- * This class is used to configure and execute CMS Business rules.
+ * This class is used to config
```
Comment (jasonaowen) [inline psm-app/cms-business-process/src/main/java/gov/medicaid/domain/rules/DroolsKnowledgeDelegate.java]: Did you intend to truncate this?

File: psm-app/cms-business-process/src/main/java/gov/medicaid/domain/rules/RulesExecutor.java [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+     * @param entityManager
+     * @param ejbContext
+     * @throws Exception
+     */
+    public static void resubmitEnrollment(CMSUser user, CMSUser systemUser,
+        EnrollmentProcess enrollmentProcess,
+        Set<Map.Entry<String, WorkItemHandler>> workItemHandlerEntrySet,
+        ProviderEnrollmentService providerEnrollmentService,
+        EntityManagerFactory entityManagerFactory, EntityManager entityManager,
+        EJBContext ejbContext) throws Exception {
```
Comment (jasonaowen) [inline psm-app/cms-business-process/src/main/java/gov/medicaid/domain/rules/RulesExecutor.java]: I super appreciate you wrapping long lines!

The style we have adopted in wrapping long lines, and which I sadly can't figure out how to encode in Checkstyle, is to put each parameter on its own line. This makes future changes to the method or its parameters (including renaming the method, renaming its parameters, changing the return type, changing the parameter type, adding parameters, or removing parameters) produce a more readable diff: it is easier to see what has changed and what hasn't. Also, as a purely stylistic choice that we've been using consistently, the closing paren, any checked exceptions, and opening brace (`) {`) goes on its own line.

See, for example, [EnrollmentController#exportBatch](https://github.com/SolutionGuidance/psm/blob/7d9981c952da98f9520988dd0a91e06a1ac45ff4/psm-app/cms-web/src/main/java/gov/medicaid/controllers/admin/EnrollmentController.java#L394-L397).

(Also, to be clear, this applies to all our method definitions and declarations; not just the one this comment is attached to!)

File: psm-app/cms-business-process/src/main/java/gov/medicaid/domain/rules/RulesExecutor.java [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+import java.util.logging.Logger;
+
+import static java.util.logging.Level.SEVERE;
+
+/**
+ * This class handles all KnowledgeSession operations.
+ */
+public class RulesExecutor {
+
+    private static final Logger logger = Logger.getLogger(RulesExecutor.class.getName());
```
[the remark this answers is not in the document]
Comment (mwolfthal) [reply psm-app/cms-business-process/src/main/java/gov/medicaid/domain/rules/RulesExecutor.java]: However at present all the methods in this class are static. That would require creating a static instance internally to hold the logger and then referencing instance.logger from each method. Is that what you want?

File: psm-app/cms-business-process/src/main/java/gov/medicaid/domain/rules/DroolsKnowledgeDelegate.java [hunk 4]
```diff
@@ -75,63 +77,91 @@
...
+        return screeningKnowledgeBase.newStatefulKnowledgeSession();
+    }
+
+
+    /**
+     * @param sessionId
+     * @param entityManagerFactory
+     * @param utx
+     * @return StatefulKnowledgeSession
+     */
```
[the remark this answers is not in the document]
Comment (mwolfthal) [reply psm-app/cms-business-process/src/main/java/gov/medicaid/domain/rules/DroolsKnowledgeDelegate.java]: My plan is to add more detail, else I'll remove them.

File: psm-app/cms-business-process/src/main/java/gov/medicaid/domain/rules/DroolsKnowledgeDelegate.java (hunk 1 above)
Comment (mwolfthal) [reply psm-app/cms-business-process/src/main/java/gov/medicaid/domain/rules/DroolsKnowledgeDelegate.java]: thanks for catching that
 * This class is used to configure and execute CMS Business rule containers.

**Comment by mwolfthal** [comment]
Rebuilding 385-drools-update

**Pull request closed without merging**